File: 백준/Gold/2573. 빙산/빙산.py
from collections import deque
import sys
input = sys.stdin.readline


def bfs(ny,nx):
    q = deque()
    q.append((ny, nx))
    used[ny][nx] = True

    while q:
        y, x = q.popleft()
        for k in range(4):
            dy = diy[k] + y
            dx = dix[k] + x
            if dy < 0 or dy >= N or dx < 0 or dx >= M:
                continue
            if arr[dy][dx] == 0:
                used[y][x] += 1

            if not used[dy][dx] and arr[dy][dx] != 0:
                q.append((dy, dx))
                used[dy][dx] = True


N,M = map(int,input().split())
arr = [list(map(int,input().split())) for _ in range(N)]
diy = [0,0,1,-1]
dix = [1,-1,0,0]
ans = 0

while 1:
    used = [[0] * M for _ in range(N)]
    ice = 0
    for i in range(N):
        for j in range(M):
            if arr[i][j] != 0 and used[i][j] == 0 :
                bfs(i, j)
                ice += 1

    for i in range(N):
        for j in range(M):
            if used[i][j]:
                arr[i][j] -= (used[i][j]-1)
            if arr[i][j] <0: arr[i][j] = 0

    ans += 1
    if ice == 0:
        print(0)
        exit()
    if ice>=2:
        break

print(ans-1)

File: 백준/Gold/2573. 빙산/빙산.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Main {
    static int N, M;
    static int[][] arr;
    static int[][] used;
    static int[] diy = {0, 0, 1, -1};
    static int[] dix = {1, -1, 0, 0};

    public static void bfs(int ny, int nx) {
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{ny, nx});
        used[ny][nx] = 1;

        while (!q.isEmpty()) {
            int[] current = q.poll();
            int y = current[0];
            int x = current[1];

            for (int k = 0; k < 4; k++) {
                int dy = diy[k] + y;
                int dx = dix[k] + x;

                if (dy < 0 || dy >= N || dx < 0 || dx >= M) {
                    continue;
                }

                if (arr[dy][dx] == 0) {
                    used[y][x]++;
                } else if (used[dy][dx] == 0) {
                    q.offer(new int[]{dy, dx});
                    used[dy][dx] = 1;
                }
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        M = sc.nextInt();
        arr = new int[N][M];

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                arr[i][j] = sc.nextInt();
            }
        }

        int ans = 0;

        while (true) {
            used = new int[N][M];
            int ice = 0;

            for (int i = 0; i < N; i++) {
                for (int j = 0; j < M; j++) {
                    if (arr[i][j] != 0 && used[i][j] == 0) {
                        bfs(i, j);
                        ice++;
                    }
                }
            }
            
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < M; j++) {
                    if (used[i][j] != 0) {
                        arr[i][j] -= (used[i][j] - 1);
                        if (arr[i][j] < 0) {
                            arr[i][j] = 0;
                        }
                    }
                }
            }

            if (ice == 0) {
                System.out.println(0);
                return;
            }
            if (ice >= 2) {
                System.out.println(ans);
                return;
            }

            ans++;
        }
    }
}